internal/healthchecks/builtin.go

```diff
@@ -0,0 +1,216 @@
+		obj, err := runtime.ObjectConvertor(s).ConvertToVersion(object.DeepCopyObject(), object.GetObjectKind().GroupVersionKind().GroupVersion())
+		if err != nil {
+			gvkErrors = appendResourceError(gvkErrors, object, err.Error())
+			continue
+		}
+		switch obj := obj.(type) {
+		case *appsv1.Deployment:
+			if err := client.Get(ctx, types.NamespacedName{
+				Name:      obj.GetName(),
+				Namespace: obj.GetNamespace(),
+			}, obj); err != nil {
+				gvkErrors = appendResourceError(gvkErrors, obj, err.Error())
+				continue
+			}
+			if obj.Status.Conditions == nil {
+				gvkErrors = appendResourceError(gvkErrors, obj, "Deployment has no conditions")
+				continue
+			}
+
+			// Check if the deployment is available.
+			conditionExists := false
+			for _, condition := range obj.Status.Conditions {
+				if condition.Type == appsv1.DeploymentAvailable {
+					if condition.Status != "True" {
+						gvkErrors = appendResourceError(gvkErrors, obj, condition.Message)
+					}
+					conditionExists = true
+					break
+				}
+			}
+			if conditionExists {
+				continue
+			}
+			gvkErrors = appendResourceError(gvkErrors, obj, "DeploymentAvailable condition not found")
+		case *appsv1.StatefulSet:
+			if err := client.Get(ctx, types.NamespacedName{
+				Name:      obj.GetName(),
+				Namespace: obj.GetNamespace(),
+			}, obj); err != nil {
+				gvkErrors = appendResourceError(gvkErrors, obj, err.Error())
+				continue
+			}
+			if obj.Status.ReadyReplicas != obj.Status.Replicas {
+				gvkErrors = appendResourceError(gvkErrors, obj, "StatefulSet is not ready")
+			}
+		case *appsv1.DaemonSet:
+			if err := client.Get(ctx, types.NamespacedName{
+				Name:      obj.GetName(),
+				Namespace: obj.GetNamespace(),
+			}, obj); err != nil {
+				gvkErrors = appendResourceError(gvkErrors, obj, err.Error())
+				continue
+			}
+			if obj.Status.NumberAvailable != obj.Status.DesiredNumberScheduled {
+				gvkErrors = appendResourceError(gvkErrors, obj, "DaemonSet is not ready")
+			}
+		case *appsv1.ReplicaSet:
+			if err := client.Get(ctx, types.NamespacedName{
+				Name:      obj.GetName(),
+				Namespace: obj.GetNamespace(),
+			}, obj); err != nil {
+				gvkErrors = appendResourceError(gvkErrors, obj, err.Error())
+				continue
+			}
+			if obj.Status.AvailableReplicas != obj.Status.Replicas {
+				gvkErrors = appendResourceError(gvkErrors, obj, "ReplicaSet is not ready")
+			}
+		case *corev1.Pod:
+			if err := client.Get(ctx, types.NamespacedName{
+				Name:      obj.GetName(),
+				Namespace: obj.GetNamespace(),
+			}, obj); err != nil {
+				gvkErrors = appendResourceError(gvkErrors, obj, err.Error())
+				continue
+			}
+			if obj.Status.Conditions == nil {
+				gvkErrors = appendResourceError(gvkErrors, obj, "Pod has no conditions")
+				continue
+			}
+			if obj.Status.Phase != corev1.PodRunning && obj.Status.Phase != corev1.PodSucceeded {
+				gvkErrors = appendResourceError(gvkErrors, obj, "Pod has no conditions")
```

Comment:
Copy/pasta typo? Seems like the error message should be different here.

Reply:
done